### PR TITLE
Add Color Transformation

### PR DESCRIPTION
**Pull request by seungjaeryanlee**
This transformation augments the input sentence by randomly replacing colors.

For example,

> I bought this **pink** shoes today! Isn't it pretty? 

becomes

> I bought this **misty rose** shoes today! Isn't it pretty?

### CODE REVIEW

File: transformations/color_transformation/README.md
Comment (kaustubhdhole) [inline transformations/color_transformation/README.md:30]: Hi @seungjaeryanlee thank you for your changes. It would be useful to provide a URL link/citation where these colors are available.

**Comment by kaustubhdhole** [comment]
I think this is a great PR! To keep the meaning intact, I would suggest one thing - since these are a handful of colors, you could make a map of color to these browser colors instead of random swapping (maybe an argument in the constructor). This would ensure that you your transformation would also retain the meaning. eg. `The sun looks yellow today.` should not be replaced with `The sun looks Alice Black today` but only only with nearby colors `The sun looks Gold/YellowGreen`. I also believe these colors could be somewhat ambiguous and context dependent so such a map based substitution might not be near-perfect. What do you think?

File: transformations/color_transformation/README.md
Comment (seungjaeryanlee) [reply transformations/color_transformation/README.md]: Thank you for your suggestion! I have added an appropriate citation.

**Comment by seungjaeryanlee** [comment]
Thank you for your review @kaustubhdhole! I agree that allowing a custom, selective mapping would add more control to the transformation. It will also indirectly disallow certain colors. I will make the edit soon.

As you have said, the mapping still has its problems, and there can be other additions to make it better. One idea might be to only detect adjectives. What do you think?

File: transformations/color_transformation/color_names.txt
Comment (Saad-Mahamood) [inline transformations/color_transformation/color_names.txt:1]: I would like to see this structured as a JSON or XML file. So that it can be easy in the future to add other language support.

File: transformations/color_transformation/test.json
Comment (Saad-Mahamood) [inline transformations/color_transformation/test.json:43]: Should be: All swan**s** **are** white.

**Review by Saad-Mahamood — CHANGES_REQUESTED** [review]
Just a few minor improvements needed before approval.

**Comment by kaustubhdhole** [comment]
@seungjaeryanlee I think keeping only adjectives should also be fine. But most color usages you would probably encounter would be adjectives itself. So, I am skeptical if restricting it with adjectives would benefit it alot (instead might add a bit of overhead). So, if you could address @Saad-Mahamood's comments and my comments about restricting the space of colors, that'll be great!

**Comment by kaustubhdhole** [comment]
@seungjaeryanlee

**Comment by seungjaeryanlee** [comment]
Oops this somehow passed my radar. Thanks for the ping! To summarize, the following needs to be done:

1. Change `color_names.txt` to `color_names.json`
2. Fix L43 of test.json
3. Allow custom mapping

I'll get this done by tomorrow.

**Comment by kaustubhdhole** [comment]
Okay, these are interesting changes - some final recommendations before merging - Please add the relevant [keywords](https://github.com/GEM-benchmark/NL-Augmenter/blob/main/docs/keywords.md) too. Also a [robustness evaluation](https://github.com/GEM-benchmark/NL-Augmenter/tree/main/evaluation) would make this PR stronger.

File: transformations/color_transformation/test.json
Comment (kaustubhdhole) [inline transformations/color_transformation/test.json:16]: * shoes --> shoe

File: transformations/color_transformation/colors.json
Comment (kaustubhdhole) [inline transformations/color_transformation/colors.json:461]: Also, if the decimal and hex values are not being used, you might as well remove them?

**Comment by seungjaeryanlee** [comment]
Thanks for the review!

I added the colors initially to add some pre-configured color-based mappings but decided against it, so I'll remove them. I'll check the list of keywords and add relevant ones: I might have few questions about them. Not sure if my machine is powerful enough for evaluation, but we'll see!

Just adding a checklist for myself:

- [ ] [Add relevant keywords](https://github.com/GEM-benchmark/NL-Augmenter/blob/main/docs/keywords.md)
- [ ] [Do robustness evaluation](https://github.com/GEM-benchmark/NL-Augmenter/tree/main/evaluation)
- [ ] Fix grammar
- [ ] Remove unused parameters

**Comment by seungjaeryanlee** [comment]
Thanks for the review!

I added the colors initially to add some pre-configured color-based mappings but decided against it, so I'll remove them. I'll check the list of keywords and add relevant ones: I might have few questions about them. Not sure if my machine is powerful enough for evaluation, but we'll see!

Just adding a checklist for myself:

- [ ] [Add relevant keywords](https://github.com/GEM-benchmark/NL-Augmenter/blob/main/docs/keywords.md)
- [ ] [Do robustness evaluation](https://github.com/GEM-benchmark/NL-Augmenter/tree/main/evaluation)
- [ ] Fix grammar
- [ ] Remove unused parameters

File: transformations/color_transformation/test.json
Comment (Saad-Mahamood) [reply transformations/color_transformation/test.json:16]: Or change this to these instead.

File: transformations/color_transformation/transformation.py
Comment (Saad-Mahamood) [inline transformations/color_transformation/transformation.py:31]: Can you add the missing [keywords](https://github.com/GEM-benchmark/NL-Augmenter/blob/main/docs/keywords.md) variable here.

**Review by Saad-Mahamood — CHANGES_REQUESTED** [review]
Only missing the required keywords variable before approval can be given.

File: transformations/color_transformation/colors.json
Comment (seungjaeryanlee) [reply transformations/color_transformation/colors.json:461]: Sounds good! I will keep the dict structure since it's probably easier to add things later.

File: transformations/color_transformation/test.json
Comment (seungjaeryanlee) [reply transformations/color_transformation/test.json:16]: Thanks for the catch! Changed it to "pair of shoes."

File: transformations/color_transformation/transformation.py
Comment (seungjaeryanlee) [reply transformations/color_transformation/transformation.py:42]: Done! Thanks for the pointer.

**Comment by seungjaeryanlee** [comment]
I made all the requested fixes. Please tell me if there is anything else to fix, or if some keywords need to be changed. Thanks!

**Review by Saad-Mahamood — APPROVED** [review]

**Review by kaustubhdhole — APPROVED** [review]

**Comment by kaustubhdhole** [comment]
Thank you for the hard work @seungjaeryanlee and thanks for the review @Saad-Mahamood

**Pull request merged**
